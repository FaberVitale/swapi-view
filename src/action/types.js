//@flow
export const FETCH_RESOURCE = "FETCH_RESOURCE";

export const FETCH_FAIL = "FETCH_FAIL";

export const FETCH_SUCCESS = "FETCH_SUCCESS";

export const REQUEST_CHANGE_DRAWER = "REQUEST_CHANGE_DRAWER";

export const SAVE_IMAGE_URL = "SAVE_IMAGE_URL";

export const LOAD_IMAGE = "LOAD_IMAGE";

export const FAILED_TO_LOAD_IMAGE = "FAILED_TO_LOAD_IMAGE";
