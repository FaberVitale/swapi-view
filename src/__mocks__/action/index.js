import {
  FETCH_RESOURCE,
  FETCH_SUCCESS,
  REQUEST_CHANGE_DRAWER,
  FETCH_FAIL
} from "../../action/types";

import { api } from "../../requestOptions";

export const fetchResourceAction = {
  page: {
    type: FETCH_RESOURCE,
    request: {
      name: "/people/page/5",
      type: "page",
      init: api,
      url: "https://swapi.co/api/people?page=5",
      query: null,
      match: {
        url: "/people/page/5",
        params: {
          endpoint: "people",
          type: "page",
          id: "5"
        }
      },
      shouldInvalidate: false
    }
  },
  item: {
    type: FETCH_RESOURCE,
    request: {
      name: "/people/item/5",
      type: "item",
      url: "https://swapi.co/api/people/5",
      init: api,
      query: null,
      match: {
        url: "/people/item/5",
        params: {
          endpoint: "people",
          type: "item",
          id: "5"
        }
      },
      shouldInvalidate: false
    }
  },
  wookiee: {
    type: FETCH_RESOURCE,
    request: {
      init: api,
      name: "/people/wookiee/5",
      type: "wookiee",
      url: "https://swapi.co/api/people/5?format=wookiee",
      query: null,
      match: {
        url: "/people/wookiee/5",
        params: {
          endpoint: "people",
          type: "wookiee",
          id: "5"
        }
      },
      shouldInvalidate: false
    }
  },
  search: {
    type: "FETCH_RESOURCE",
    request: {
      init: api,
      name: "/people/search/abba/1",
      url: "https://swapi.co/api/people?search=abba&page=1",
      query: {
        q: "abba",
        page: "1"
      },
      match: {
        path:
          "/:endpoint(people|planets|films|species|vehicles|starships)/:type(search)",
        url: "/people/search",
        isExact: true,
        params: {
          endpoint: "people",
          type: "search"
        }
      },
      type: "search",
      shouldInvalidate: false
    }
  }
};

export const fetchFailAction = {
  type: FETCH_FAIL,
  request: {
    name: "/people/page/5"
  },
  cause: "500"
};

export const fetchSuccessAction = {
  page: {
    type: FETCH_SUCCESS,
    request: {
      init: api,
      type: "page",
      name: "/people/page/5",
      query: null,
      match: {
        url: "/people/page/5",
        params: {
          endpoint: "people",
          type: "page",
          id: "5"
        }
      }
    },
    data: {
      previous: "/people/page/4",
      next: "/people/page/6",
      ids: ["/people/item/1", "/people/item/2"],
      results: [
        { data: {}, meta: { url: "1" }, links: {} },
        { data: {}, meta: { url: "2" }, links: {} }
      ]
    }
  },

  item: {
    type: FETCH_SUCCESS,
    request: {
      init: api,
      type: "item",
      name: "people/item/2",
      query: null,
      match: {
        url: "people/item/2",
        params: {
          endpoint: "people",
          type: "item",
          id: "2"
        }
      }
    },
    data: { data: {}, meta: {}, links: {} }
  },

  wookiee: {
    type: FETCH_SUCCESS,
    request: {
      init: api,
      type: "wookiee",
      name: "people/wookiee/2",
      query: null,
      match: {
        url: "people/wookiee/2",
        params: {
          endpoint: "people",
          type: "wookiee",
          id: "2"
        }
      }
    },
    data: { data: {}, meta: {}, links: {} }
  },
  search: {
    type: "FETCH_SUCCESS",
    request: {
      init: api,
      name: "/people/search/luke/1",
      url: "https://swapi.co/api/people/?search=luke&page=1",
      query: {
        q: "luke",
        page: "1"
      },
      match: {
        path:
          "/:endpoint(people|planets|films|species|vehicles|starships)/:type(search)",
        url: "/people/search",
        isExact: true,
        params: {
          endpoint: "people",
          type: "search"
        }
      },
      type: "search",
      shouldInvalidate: false
    },
    data: {
      next: null,
      previous: null,
      ids: ["/people/item/1"],
      results: [
        {
          meta: {
            created: "2014-12-09T13:50:51.644000Z",
            edited: "2014-12-20T21:17:56.891000Z",
            url: "https://swapi.co/api/people/1/"
          },
          data: {
            name: "Luke Skywalker",
            height: "172",
            mass: "77",
            hair_color: "blond",
            skin_color: "fair",
            eye_color: "blue",
            birth_year: "19BBY",
            gender: "male"
          },
          links: {
            homeworld: "https://swapi.co/api/planets/1/",
            films: [
              "https://swapi.co/api/films/2/",
              "https://swapi.co/api/films/6/",
              "https://swapi.co/api/films/3/",
              "https://swapi.co/api/films/1/",
              "https://swapi.co/api/films/7/"
            ],
            species: ["https://swapi.co/api/species/1/"],
            vehicles: [
              "https://swapi.co/api/vehicles/14/",
              "https://swapi.co/api/vehicles/30/"
            ],
            starships: [
              "https://swapi.co/api/starships/12/",
              "https://swapi.co/api/starships/22/"
            ]
          }
        }
      ]
    }
  }
};

export const drawerAction = {
  type: REQUEST_CHANGE_DRAWER,
  value: true
};
