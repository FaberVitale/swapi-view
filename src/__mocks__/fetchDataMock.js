//@flow
import item from "../__mocks__/api/item";
import page from "../__mocks__/api/page";
import wookiee from "../__mocks__/api/wookiee";
import search from "../__mocks__/api/search";
import emptySearch from "../__mocks__/api/emptySearch";
/* we use this async function to do offline testing 
 * given a type requested it always returns an object
 *  with the same type taken from __mock__/api
 */

export const fetchDataMock = (timeout: number = 10000) => {
  return async (url: string) => {
    if (/\?search/.test(url)) {
      if (url.indexOf("empty") > -1) {
        return emptySearch;
      }
      return search;
    }

    if (url.indexOf("wookiee") > -1) {
      return wookiee;
    } else if (url.indexOf("page") > -1) {
      return page;
    }

    return item;
  };
};
