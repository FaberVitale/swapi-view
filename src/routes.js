//@flow
import qs from "qs";

export const apiRoot = "https://swapi.co/api/";
export const endpoints = [
  "people",
  "planets",
  "films",
  "species",
  "vehicles",
  "starships"
];
export const firstPageOf = {
  people: "/people/page/1",
  planets: "/planets/page/1",
  films: "/films/page/1",
  species: "/species/page/1",
  vehicles: "/vehicles/page/1",
  starships: "/starships/page/1"
};
export const paths = {
  pages:
    "/:endpoint(people|planets|films|species|vehicles|starships)/:type(page)/:id([1-9]\\d?)",
  search:
    "/:endpoint(people|planets|films|species|vehicles|starships)/:type(search)",
  item:
    "/:endpoint(people|planets|films|species|vehicles|starships)/:type(item)/:id([1-9]\\d?)",
  wookiee:
    "/:endpoint(people|planets|films|species|vehicles|starships)/:type(wookiee)/:id([1-9]\\d?)",
  about: "/about",
  home: "/",
  fileNotFound: "/404"
};

export const qsOpt = {
  format: qs.formats.RFC1738
};

export const queryRegex = {
  q: /^.{2,45}$/,
  page: /^[1-9]$/
};

export const searchFieldsOf = {
  people: ["name"],
  planets: ["name"],
  films: ["title"],
  species: ["name"],
  vehicles: ["name", "model"],
  starships: ["name", "model"]
};
