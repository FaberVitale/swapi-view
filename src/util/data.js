//@flow
const isMetaKey = (key: string) =>
  /^(?:created|url|edited|oarcworaaowowa|wowaahaowowa|hurcan)$/.test(key);
const isLink = value =>
  typeof value !== "string" || /^https?|acaoaoakc/.test(value);

const regex = /(\d+)\/?$/;
export const mapTypeToRoute = (
  type: string,
  endpoint: string,
  apiUrl: string,
  query?: { q: string, page: string } | null
) => {
  const res = regex.exec(apiUrl);

  if (type === "search" && query && res) {
    return `/${endpoint}/search/${query.q}/${res[1]}`;
  }

  return res && `/${endpoint}/${type}/${res[1]}`;
};

export const transformItem = (item: {}) =>
  Object.keys(item || {}).reduce(
    (aggr, key) => {
      const value = item[key];

      if (isMetaKey(key)) {
        aggr.meta[key] = value;
      } else if (isLink(value)) {
        aggr.links[key] = value;
      } else {
        aggr.data[key] = value;
      }
      return aggr;
    },
    { meta: {}, data: {}, links: {} }
  );

export const transformPage = (
  request: FetchRequest,
  page: {
    next: string | null,
    previous: string | null,
    results: Array<{ url: string }>
  }
) => {
  const type = request.type;
  const { endpoint } = request.match.params;

  return {
    next: page.next && mapTypeToRoute(type, endpoint, page.next, request.query),
    previous:
      page.previous &&
      mapTypeToRoute(type, endpoint, page.previous, request.query),
    ids: page.results.map(item => mapTypeToRoute("item", endpoint, item.url)),
    results: page.results.map(transformItem)
  };
};
