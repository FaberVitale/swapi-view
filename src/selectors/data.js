//@flow
import type { State } from "../reducers";
import type { Match } from "react-router-dom";
import { createSelector } from "reselect";
import { endpoints, searchFieldsOf } from "../routes";

export const getData = (state: State) => state.data;

const getEndpointData = (state: State, match: Match) => {
  const data = getData(state);
  const endpoint = match.params.endpoint || null;

  return data && endpoint ? data[endpoint] : null;
};

export const hasResource = (state: State, request: FetchRequest) => {
  const { match, type, query, name } = request;

  switch (type) {
    case "search":
      return !!getSearchPage(state, name, query);
    case "item":
      return !!getItem(state, match);
    case "page":
      return !!getPage(state, match);
    case "wookiee":
      return !!getWookiee(state, match);
    default:
      return false;
  }
};

const getItem = (state: State, match: Match) => {
  const endpointData = getEndpointData(state, match);

  return endpointData && endpointData.items[match.url];
};

export const getPage = (state: State, match: Match) => {
  const endpointData = getEndpointData(state, match);

  return endpointData && endpointData.pages[match.url];
};

const getWookiee = (state: State, match: Match) => {
  const endpointData = getEndpointData(state, match);

  return endpointData && endpointData.wookiee[match.url];
};

const defaultSearchPage = {
  previous: null,
  next: null,
  items: []
};

export const getSearchPage = (
  state: State,
  name: string,
  query: { q: string, page: string } | null
) => {
  if (!query) {
    return defaultSearchPage;
  }

  return getData(state).search[name] || null;
};

export const getItemOrWookiee = (state: State, match: Match) => {
  return match.params.type === "item"
    ? getItem(state, match)
    : getWookiee(state, match);
};

/* getItemsOf[endpoint]:  (data) => data[endpoint].items
 * in order to work, it requires that the input is the result of getData
 * */
export const getItemsOf = endpoints.reduce((res, endpoint) => {
  res[endpoint] = data => data[endpoint].items;

  return res;
}, {});

/* Object that maps endpoints to  a function that creates an array of suggestions
 * given the endpoint given the result of getData  */
export const getSearchSuggestionsOf = (function buildSearchSuggestionsOf(
  endpoints
) {
  const retrieveSuggestions = (endpoint: string) => items => {
    let suggestions: Array<string> = [];
    const keys: Array<string> = Object.keys(items);

    const suggestionToKey = {};

    for (const key of keys) {
      const searchValues = searchFieldsOf[endpoint].map(
        searchValue => items[key].data[searchValue]
      );
      suggestions.push(...searchValues);

      for (const searchValue of searchValues) {
        suggestionToKey[searchValue] = key;
      }
    }

    // a lot vehicles and starships have duplicated values
    if (suggestions.length > 1 && searchFieldsOf[endpoint].length > 1) {
      suggestions = [...new Set(suggestions)];
    }

    return {
      suggestions,
      suggestionToKey
    };
  };

  let res: {
    [endpoint: string]: (items: {}) => {
      suggestion: Array<string>,
      suggestionToKey: { [sugg: string]: string }
    }
  } = {};

  for (const endpoint of endpoints) {
    res[endpoint] = createSelector(
      getItemsOf[endpoint],
      retrieveSuggestions(endpoint)
    );
  }

  return res;
})(endpoints || []);

/* given (state: State) passed returns an object that maps endoints to suggestions list */
export const getDataSources = createSelector(getData, function buildDatasources(
  data
) {
  const res = {};

  for (let i = 0, len = endpoints.length; i < len; i++) {
    res[endpoints[i]] = getSearchSuggestionsOf[endpoints[i]](data);
  }

  return res;
});
