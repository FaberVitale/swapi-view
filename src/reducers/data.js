//@flow
import { equalsTo, equals } from "../util/functions";
import { combineReducers, compose } from "redux";
import { filterActions } from "redux-ignore";
import pagination from "./pagination";
import byId from "./byId";
import type { State as ById } from "./byId";
import type { State as PageState } from "./pagination";
import { endpoints } from "../routes";
import { getType, getEndpoint } from "../selectors/router";

type EndpointData = {|
  wookiee: ById,
  items: ById,
  pages: PageState
|};

type MakeEndpointReducer = string => (
  state: EndpointData,
  action: Action
) => EndpointData;

type MakeDataReducer = (
  endpoints: Array<string>
) => {
  [endpoint: string]: (state: EndpointData, action: Action) => EndpointData
};

export type State = {
  [endpoint: string]: EndpointData,
  search: PageState
};

export default (function buildDataReducer(endpoints) {
  const filterFor = {
    page: compose(equals("page"), getType),
    item: compose(equalsTo("page", "item", "search"), getType),
    wookiee: compose(equals("wookiee"), getType),
    search: compose(equals("search"), getType)
  };

  const makeEndpointReducer: MakeEndpointReducer = (endpoint: string) =>
    filterActions(
      combineReducers({
        pages: filterActions(pagination, filterFor.page),
        items: filterActions(byId, filterFor.item),
        wookiee: filterActions(byId, filterFor.wookiee)
      }),
      compose(equals(endpoint), getEndpoint)
    );

  const makeDataReducer: MakeDataReducer = endpoints => {
    const res = endpoints.reduce((aggr, next) => {
      aggr[next] = makeEndpointReducer(next);

      return aggr;
    }, {});

    res.search = filterActions(pagination, filterFor.search);

    return res;
  };

  return combineReducers(makeDataReducer(endpoints));
})(endpoints || []);
