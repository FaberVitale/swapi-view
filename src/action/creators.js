//@flow
import {
  FETCH_RESOURCE,
  FETCH_SUCCESS,
  FETCH_FAIL,
  REQUEST_CHANGE_DRAWER,
  SAVE_IMAGE_URL,
  FAILED_TO_LOAD_IMAGE,
  LOAD_IMAGE
} from "./types";
import { apiRoot, qsOpt } from "../routes";
import qs from "qs";
import { getQuery } from "../util/query";
import { transformPage, transformItem, mapTypeToRoute } from "../util/data";
import { api } from "../requestOptions";

export const fetchRouteResource: FetchRouteResource = (
  match,
  search,
  shouldInvalidate
) => {
  let name: string = match.url;
  let url: string | null = null;
  const { endpoint, id, type } = match.params;
  let query = null;
  if (type === "search") {
    query = getQuery(search);

    if (query) {
      const searchString = qs.stringify(
        { search: query.q, page: query.page },
        qsOpt
      );
      url = `${apiRoot}${endpoint}?${searchString}`;
      name = mapTypeToRoute(type, endpoint, url, query);
    } else {
      name = `/${endpoint}/search/invalid_query`;
    }
  } else if (endpoint && id) {
    /* eslint-disable default-case */
    switch (type) {
      case "page":
        url = `${apiRoot}${endpoint}?page=${id}`;
        break;
      case "item":
        url = `${apiRoot}${endpoint}/${id}`;
        break;
      case "wookiee":
        url = `${apiRoot}${endpoint}/${id}?format=wookiee`;
        break;
    }
    /* eslint-enable default-case */
  }

  return {
    type: FETCH_RESOURCE,
    request: {
      name,
      url,
      init: api,
      query,
      match,
      type,
      shouldInvalidate
    }
  };
};

export const fetchSuccess: FetchSuccess = (request, data) => {
  let nextData;

  switch (request.type) {
    case "page":
      nextData = transformPage(request, data);
      break;
    case "search":
      nextData = transformPage(request, data);
      break;
    case "item":
      nextData = transformItem(data);
      break;
    case "wookiee":
      nextData = transformItem(data);
      break;
    default:
      nextData = data;
  }

  return {
    type: FETCH_SUCCESS,
    request,
    data: nextData
  };
};

export const fetchFail: FetchFail = (request, cause) => ({
  type: FETCH_FAIL,
  request,
  cause
});

export const drawerRequestChange: ChangeStateDrawer = (value: boolean) => ({
  type: REQUEST_CHANGE_DRAWER,
  value
});

export const saveImageURL: SaveImage = (blobUrl: string) => ({
  type: SAVE_IMAGE_URL,
  url: blobUrl
});

export const failToLoadImage: FailedToLoadImage = () => ({
  type: FAILED_TO_LOAD_IMAGE
});

export const loadImage: LoadImage = () => ({
  type: LOAD_IMAGE
});
