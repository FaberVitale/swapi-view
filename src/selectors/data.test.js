import * as fn from "./data";
import reducer from "../reducers";
import { fetchSuccessAction as action } from "../__mocks__/action";
import { endpoints } from "../routes";

describe("selectors/data", () => {
  const initState = reducer(undefined, {});
  const itemMatch = action.item.request.match;
  const wookieeMatch = action.wookiee.request.match;
  const pageMatch = action.page.request.match;
  test("getData", () => {
    expect(fn.getData(initState)).toBe(initState.data);
  });

  test("getPage", () => {
    expect(
      fn.getPage(reducer(initState, action.page), pageMatch)
    ).toMatchSnapshot();

    expect(fn.getPage(initState, pageMatch)).toBe(undefined);
  });

  test("getItemOrWookiee", () => {
    expect(
      fn.getItemOrWookiee(reducer(initState, action.item), itemMatch)
    ).toMatchSnapshot();
    expect(
      fn.getItemOrWookiee(reducer(initState, action.wookiee), wookieeMatch)
    ).toMatchSnapshot();

    expect(fn.getItemOrWookiee(initState, wookieeMatch)).toBe(undefined);
    expect(fn.getItemOrWookiee(initState, itemMatch)).toBe(undefined);
  });

  test("hasResource", () => {
    expect(fn.hasResource(initState, action.page.request)).toBe(false);
    expect(
      fn.hasResource(reducer(initState, action.page), action.page.request)
    ).toBe(true);
    expect(
      fn.hasResource(reducer(initState, action.item), action.item.request)
    ).toBe(true);
    expect(
      fn.hasResource(reducer(initState, action.wookiee), action.wookiee.request)
    ).toBe(true);

    expect(
      fn.hasResource(reducer(initState, action.search), action.search.request)
    ).toBe(true);
  });

  describe("getDataSources", () => {
    it("returns an object that maps endpoints to [] if the initState is passed", () => {
      const res = fn.getDataSources(initState);
      const base = {
        suggestions: [],
        suggestionToKey: {}
      };
      const expected = endpoints.reduce((obj, endpoint) => {
        obj[endpoint] = base;
        return obj;
      }, {});

      expect(res).toEqual(expected);
    });
  });

  it("returns the same object if the state.data doesnt change", () => {
    const before = fn.getDataSources(initState);
    expect(before).toBe(fn.getDataSources(initState));

    const nextState = reducer(initState, action.item);
    const after = fn.getDataSources(nextState);

    expect(before).not.toBe(after);

    expect(after).toBe(fn.getDataSources(nextState));
  });
});
