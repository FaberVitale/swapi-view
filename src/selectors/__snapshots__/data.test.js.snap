// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors/data getItemOrWookiee 1`] = `
Object {
  "data": Object {},
  "links": Object {},
  "meta": Object {},
}
`;

exports[`selectors/data getItemOrWookiee 2`] = `
Object {
  "data": Object {},
  "links": Object {},
  "meta": Object {},
}
`;

exports[`selectors/data getPage 1`] = `
Object {
  "items": Array [
    "/people/item/1",
    "/people/item/2",
  ],
  "next": "/people/page/6",
  "previous": "/people/page/4",
}
`;
